x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/apikey/CrossClusterApiKeyRoleDescriptorBuilder.java

```diff
@@ -0,0 +1,128 @@
+    private static final String[] CCR_INDICES_PRIVILEGE_NAMES = { "cross_cluster_replication", "cross_cluster_replication_internal" };
+    private static final String ROLE_DESCRIPTOR_NAME = "cross_cluster";
+
+    @SuppressWarnings("unchecked")
+    public static final ConstructingObjectParser<CrossClusterApiKeyRoleDescriptorBuilder, Void> PARSER = new ConstructingObjectParser<>(
+        "cross_cluster_api_key_request_access",
+        false,
+        (args, v) -> new CrossClusterApiKeyRoleDescriptorBuilder(
+            (List<RoleDescriptor.IndicesPrivileges>) args[0],
+            (List<RoleDescriptor.IndicesPrivileges>) args[1]
+        )
+    );
+
+    static {
+        PARSER.declareObjectArray(
+            optionalConstructorArg(),
+            (p, c) -> RoleDescriptor.parseIndexWithPrivileges(ROLE_DESCRIPTOR_NAME, CCS_INDICES_PRIVILEGE_NAMES, p),
+            new ParseField("search")
+        );
+        PARSER.declareObjectArray(
+            optionalConstructorArg(),
+            (p, c) -> RoleDescriptor.parseIndexWithPrivileges(ROLE_DESCRIPTOR_NAME, CCR_INDICES_PRIVILEGE_NAMES, p),
+            new ParseField("replication")
+        );
+    }
+
+    private final List<RoleDescriptor.IndicesPrivileges> search;
+    private final List<RoleDescriptor.IndicesPrivileges> replication;
+
+    private CrossClusterApiKeyRoleDescriptorBuilder(
+        List<RoleDescriptor.IndicesPrivileges> search,
+        List<RoleDescriptor.IndicesPrivileges> replication
+    ) {
+        this.search = search == null ? List.of() : search;
+        this.replication = replication == null ? List.of() : replication;
+    }
+
+    public RoleDescriptor build() {
+        final String[] clusterPrivileges;
+        if (search.isEmpty() && replication.isEmpty()) {
+            throw new IllegalArgumentException("must specify non-empty access for either [search] or [replication]");
+        } else if (search.isEmpty()) {
+            clusterPrivileges = CCR_CLUSTER_PRIVILEGE_NAMES;
+        } else if (replication.isEmpty()) {
+            clusterPrivileges = CCS_CLUSTER_PRIVILEGE_NAMES;
+        } else {
+            clusterPrivileges = CCS_AND_CCR_CLUSTER_PRIVILEGE_NAMES;
+        }
+
+        if (replication.stream().anyMatch(RoleDescriptor.IndicesPrivileges::isUsingDocumentOrFieldLevelSecurity)) {
+            throw new IllegalArgumentException("replication does not support document or field level security");
+        }
+
+        return new RoleDescriptor(
+            ROLE_DESCRIPTOR_NAME,
+            clusterPrivileges,
+            CollectionUtils.concatLists(search, replication).toArray(RoleDescriptor.IndicesPrivileges[]::new),
+            null
+        );
+    }
+
+    static void validate(RoleDescriptor roleDescriptor) {
+        if (false == ROLE_DESCRIPTOR_NAME.equals(roleDescriptor.getName())) {
+            throw new IllegalArgumentException("invalid role descriptor name [" + roleDescriptor.getName() + "]");
```

Comment:
I'm wondering if these should be assertions instead; we pick names, what types of privileges are defined, etc. so if these don't match what we expect it's a programming error not a user input problem.

Reply:
This method is effectively an assertion because it runs with a `if (Assertions.ENABLED)` check in `CreateCrossClusterApiKeyRequest`. I could change all throwing to assert as well. But it seems existing code mostly write such method in a regular way when its usage is wrapped inside `if (Assertions.ENABLED)`.